# fusions_mutect

This workflow analyzes WGS data to look for a short list of known fusion driver mutations of clonal hematopoiesis. It also runs Mutect2, goleft/indexcov, and mosdepth on DNMT3A regions.

**Inputs**

* `wf_fusions_bed`: fusions.py region file with breakpoint cluster regions of interest in BED format
* `wf_full_cram`: full WGS CRAM file
* `wf_full_cram_crai`: full WGS CRAM index file
* `wf_ROIs`: ROIs file with gene fusions of interest listed in gene1_gene2 format
* `wf_subset_bed`: BED file of genes of interest for Mutect2 and coverage analysis - in this primary analysis, exons only
* `wf_gene_ref_bed`: BED file for gene reference with list of all genes with positions, names, and strandedness
* `wf_ref_fa`: reference FASTA file
* `wf_ref_fa_fai`: reference FASTA index file
* `wf_ref_dict`: reference DICT file
* `Mutect2_intervals`: interval list for Mutect2 analysis, generated by picard.jar (can use BedToIntervalList function on the subset.bed file)
* `Mutect2_scatter_count`: scatter count for larger analyses, refer to Mutect2 documentation
* `Mutect2_gatk_docker`: docker image for Mutect2, refer to Mutect2 documentation
* `Mutect2_gatk_override`: path in image of gatk.jar, refer to Mutect2 documentation - in this analysis, uploaded gatk.jar to bucket and input that way due to docker image inconsistencies
* `Mutect2_run_funcotator`: whether to run Mutect2 functional annotator, refer to Mutect2 documentation
* `wf_subset_mosdepth`: BED file of regions of interest for mosdepth coverage analysis - in this case, genes + 1x gene width on either side
